### ISSUE TITLE
Can't bind to public ip address

### ISSUE THREAD

**Issue by p1ratrulezzz** [TelegramMessenger/MTProxy#2]
```
[7864][2018-05-30 18:33:01.612073 local] Invoking engine mtproxy-0.01 compiled at May 30 2018 18:06:31 by gcc 7.3.0 64-bit after commit 2fd27cb0753c62e5766c86db28f2acc1d755547f
[7864][2018-05-30 18:33:01.612530 local] config_filename = 'proxy-multi.conf'
[7864][2018-05-30 18:33:01.613037 local] creating 1 workers
[7864][2018-05-30 18:33:01.615300 local] Bad binded IP address 95.123.456.78, search in ifconfig
[7864][2018-05-30 18:33:01.616752 local] Started as [10.114.26.50:8888:7864:1527697981]
[7864][2018-05-30 18:33:01.617637 local] configuration file proxy-multi.conf re-read successfully (797 bytes parsed), new configuration active
[7864][2018-05-30 18:33:01.619576 local] main loop
[7865][2018-05-30 18:33:01.620906 local] Bad binded IP address 95.123.456.78, search in ifconfig
[7865][2018-05-30 18:33:01.622217 local] Started as [10.114.26.50:8888:7865:1527697981]
[7865][2018-05-30 18:33:01.623401 local] configuration file proxy-multi.conf re-read successfully (797 bytes parsed), new configuration active
[7865][2018-05-30 18:33:01.626701 local] main loop

```

So i have VPN (openvpn) running on tun0 interface. And the main interface is eth0. I'm trying to set public ip address to listen on it  (--address 95.123.456.78) and have this message

**Comment by p1ratrulezzz** [comment]
Solved. Just commented out the line with

```
ipv4=0;
```

on https://github.com/TelegramMessenger/MTProxy/blob/master/engine/engine.c line 230

```
if ((ipv4 >> 24) != 10) {
      kprintf ("Bad binded IP address " IP_PRINT_STR ", search in ifconfig\n", IP_TO_PRINT (ipv4));
      kprintf("But I dont care about it...");
      //ipv4 = 0;
    } 
```

Server started and serving connections now

**Comment by stek29** [comment]
It probably should’ve been == 10, since 10.0.0.0/8 is a private network.
However, rfc1918 specifies two more subnets for private networks which are not handled.

get_my_ipv4 also only skips 10.0.0.0/8 and loopback

https://github.com/TelegramMessenger/MTProxy/blob/cc7b7097a5bc1681d0f4c05d5f07576c8b279fc4/net/net-events.c#L737

**Comment by p1ratrulezzz** [comment]
Just disable this ip address check on line 230 as i mentioned in comment 2 (temporary, until developer fixes it)

**Comment by zvpdev** [comment]
Same issue.
Change from `ipv4 = 0;` to `//ipv4 = 0;` doesn't help me

**Comment by p1ratrulezzz** [comment]
@zvpdev  forcely bind to public address by passing 

--address 11.22.33.44

**Comment by p1ratrulezzz** [comment]
Also bug label needed for this, as it determines correct ip addresses as incorrect.

**Comment by p1ratrulezzz** [comment]
I'm really not sure what does this check do but as @stek29 noticed it checks only networks like 10.10... But rfc has two more ranges so I've added them too. Binding to external ip actually work so

**Comment by liquidpredator** [comment]
Same issue.
Change from ipv4 = 0; to //ipv4 = 0 and use --address 
binding to local ip anyway...

**Comment by p1ratrulezzz** [comment]
@liquidpredator try cloning my fork of this proxy. you should use --address ur.if.ip.adr for example --address 8.8.8.8 (check interface's ip address in ifconfig)

**Comment by L11R** [comment]
@p1ratrulezzz after last update with padding support seems like it broken again. Is there new workaround?

**Comment by p1ratrulezzz** [comment]
Yes, i also noticed this. I'm using my fork with older version, it works
for me so don't see any reason to upgrade

вт, 10 июл. 2018 г., 22:39 Savely Krasovsky <notifications@github.com>:

**Comment by L11R** [comment]
@p1ratrulezzz I heard Rostelecom already testing DPI blocking by packet size. So consider to update and stuck...

**Comment by JaminZhou** [comment]
`./mtproto-proxy -u nobody -p 8888 -H 443 -S <secret> --nat-info <intranet ip>:<public ip> --aes-pwd proxy-secret proxy-multi.conf -M 1
`

`--nat-info <intranet ip>:<public ip> ` add this will work

**Comment by mssucksalot** [comment]
> `./mtproto-proxy -u nobody -p 8888 -H 443 -S <secret> --nat-info <intranet ip>:<public ip> --aes-pwd proxy-secret proxy-multi.conf -M 1 `
> 
> `--nat-info <intranet ip>:<public ip> ` add this will work

this will not work if you want to use white static IP, and we still disable check, if address binding is used

**Comment by mssucksalot** [comment]
> Solved. Just commented out the line with
> 
> ```
> ipv4=0;
> ```
> 
> on https://github.com/TelegramMessenger/MTProxy/blob/master/engine/engine.c line 230
> 
> ```
> if ((ipv4 >> 24) != 10) {
>       kprintf ("Bad binded IP address " IP_PRINT_STR ", search in ifconfig\n", IP_TO_PRINT (ipv4));
>       kprintf("But I dont care about it...");
>       //ipv4 = 0;
>     } 
> ```
> 
> Server started and serving connections now. So looks like there is some issue with ipv4 check

does not work with public IP, after fix message is disappeared but proxy does not work

**UPDATE**

After this update systemd unit cannot start on server boot with "Cannot assign requested address". Though manual restart works now. And all that can be solved by systemd sleep target.

**Comment by vladsf** [comment]
Not fixed yet.
